tyro/_cli.py

```diff
@@ -241,18 +293,26 @@ def _cli_impl(
 
     # Read and fix arguments. If the user passes in --field_name instead of
     # --field-name, correct for them.
-    args = sys.argv[1:] if args is None else args
-
-    def fix_arg(arg: str) -> str:
-        if not arg.startswith("--"):
-            return arg
-        if "=" in arg:
-            arg, _, val = arg.partition("=")
-            return arg.replace("_", "-") + "=" + val
-        else:
-            return arg.replace("_", "-")
-
-    args = list(map(fix_arg, args))
+    args = list(sys.argv[1:]) if args is None else list(args)
+
+    # Fix arguments. This will modify all option-style arguments replacing
+    # underscores with dashes. This is to support the common convention of using
+    # underscores in variable names, but dashes in command line arguments.
+    # If two options are ambiguous, e.g., --a_b and --a-b, raise a runtime error.
+    modified_args: Dict[str, str] = {}
+    for index, arg in enumerate(args):
+        if arg.startswith("--"):
+            if "=" in arg:
+                arg, _, val = arg.partition("=")
+                fixed = arg.replace("_", "-") + "=" + val
+            else:
+                fixed = arg.replace("_", "-")
+            if fixed in modified_args:
```

Comment:
Would it make sense to make this condition something like:
```python
if (
    return_unknown_args
    and fixed in modified_args
    and modified_args[fixed] != arg
):
```

The main motivations are:
- To avoid an unnecessary error if we don't need to "un-fix" args.
- To be OK with duplicate argument names, like writing `--arg-one` or `--arg_one` twice as long as they're consistent. I can think of this happening if we have `action="append"` for some argument in argparse, or for the same argument existing in multiple subcommands.
    - A test would be nice for this duplicate argument thing + `parse_unknown_args()`

Reply:
Hmm, I'd be hesitant to implement different parsing logic depending on `return_unknown_args`. I do think having consistent option strings should be supported.

I added this if statement as you described it and also added a test for consistent options as well as a test for the different behaviours depending on the value of `return_unknown_args`.

Reply:
Thanks!

> Hmm, I'd be hesitant to implement different parsing logic depending on `return_unknown_args`. I do think having consistent option strings should be supported.

I agree with this in principle, although this particular case seems minor enough that I'm with erring towards being tolerant. Don't feel strongly at all though.